service/index/writer.go

```diff
@@ -15,58 +15,77 @@
 package index
 
 import (
+	"context"
+	"errors"
 	"fmt"
+	"sync"
 
 	"github.com/dgraph-io/badger/v2"
-	"github.com/optakt/flow-dps/models/dps"
+	"golang.org/x/sync/semaphore"
 
 	"github.com/onflow/flow-go/ledger"
 	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/models/dps"
 )
 
 // Writer implements the `index.Writer` interface to write indexing data to
 // an underlying Badger database.
 type Writer struct {
-	db  *badger.DB
-	lib dps.WriteLibrary
+	sync.RWMutex
+	db   *badger.DB
+	lib  dps.WriteLibrary
+	cfg  Config
+	tx   *badger.Txn
+	sema *semaphore.Weighted
+	err  error
```

Comment:
The best practice IMO when it comes to locks is that it should be obvious by their definition what exactly they protect.

A mutex placed at the top of a struct by composition implies to me that it protects all attributes and that none of them can be read/written without locking.

Since here it seems like we are only protecting the `err` field with the mutex though, I would change this to be:

```suggestion
	db   *badger.DB
	lib  dps.WriteLibrary
	cfg  Config
	tx   *badger.Txn
	sema *semaphore.Weighted
	
	
	errMu sync.RWMutex
	err  error
```

For the semaphore it's a bit different though, since it does not really protect anything, so I'm fine with the way it is used currently.

Reply:
I think you are probably right about your observation regarding explicitness in this case. At the same time, I'm reluctant to implement it, as it will increase the code verbosity.

My general practice about locks is that when a component that needs to guard some operations against concurrent access, I just place a global lock on that component, which can be used globally (like the synchronized keyword in Java). Once the component needs to guard different things against concurrent access, it needs to become explicit.